File: NotePal/Windows/NotePal.Windows/Services/EncryptedThemeAsset.cs
using System.Security.Cryptography;
using System.Text;

namespace NotePalWindows.Services;

public static class EncryptedThemeAsset
{
    public const string FileExtension = "notepalasset";

    private static readonly byte[] Magic = Encoding.UTF8.GetBytes("NPETASSET1");
    // Keep the original associated data so existing encrypted theme assets remain readable.
    private static readonly byte[] AssociatedData = Encoding.UTF8.GetBytes("NotePet encrypted theme asset v1");
    private const int SaltByteCount = 16;
    private const int NonceByteCount = 12;
    private const int TagByteCount = 16;
    private const int KeyByteCount = 32;
    private const int KeyDerivationRounds = 180_000;

    public static string NormalizeKey(string key)
    {
        return key.Trim().ToLowerInvariant();
    }

    public static byte[] Decrypt(byte[] encryptedData, string rawKey)
    {
        var keyText = NormalizeKey(rawKey);
        if (keyText.Length == 0)
        {
            throw new InvalidOperationException("Theme key is empty.");
        }

        var minimumByteCount = Magic.Length + SaltByteCount + NonceByteCount + TagByteCount + 1;
        if (encryptedData.Length < minimumByteCount || !encryptedData.AsSpan(0, Magic.Length).SequenceEqual(Magic))
        {
            throw new InvalidOperationException("Theme asset is invalid.");
        }

        var offset = Magic.Length;
        var salt = encryptedData.AsSpan(offset, SaltByteCount).ToArray();
        offset += SaltByteCount;
        var nonce = encryptedData.AsSpan(offset, NonceByteCount).ToArray();
        offset += NonceByteCount;

        var sealedData = encryptedData.AsSpan(offset).ToArray();
        if (sealedData.Length <= TagByteCount)
        {
            throw new InvalidOperationException("Theme asset is invalid.");
        }

        var cipherTextLength = sealedData.Length - TagByteCount;
        var ciphertext = sealedData.AsSpan(0, cipherTextLength).ToArray();
        var tag = sealedData.AsSpan(cipherTextLength, TagByteCount).ToArray();
        var plaintext = new byte[ciphertext.Length];
        var key = Rfc2898DeriveBytes.Pbkdf2(
            Encoding.UTF8.GetBytes(keyText),
            salt,
            KeyDerivationRounds,
            HashAlgorithmName.SHA256,
            KeyByteCount);

        using var aes = new AesGcm(key, TagByteCount);
        aes.Decrypt(nonce, ciphertext, tag, plaintext, AssociatedData);
        return plaintext;
    }
}

File: NotePal/Sources/NotePal/App/EncryptedThemeAsset.swift
import CommonCrypto
import CryptoKit
import Foundation

enum EncryptedThemeAsset {
    static let fileExtension = "notepalasset"

    private static let magic = Data("NPETASSET1".utf8)
    // Keep the original associated data so existing encrypted theme assets remain readable.
    private static let associatedData = Data("NotePet encrypted theme asset v1".utf8)
    private static let saltByteCount = 16
    private static let nonceByteCount = 12
    private static let tagByteCount = 16
    private static let keyByteCount = 32
    private static let keyDerivationRounds: UInt32 = 180_000

    static func normalizedKey(_ key: String) -> String {
        key.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    static func decrypt(_ encryptedData: Data, with rawKey: String) throws -> Data {
        let keyText = normalizedKey(rawKey)
        guard !keyText.isEmpty else {
            throw EncryptedThemeAssetError.emptyKey
        }

        let minimumByteCount = magic.count + saltByteCount + nonceByteCount + tagByteCount + 1
        guard encryptedData.count >= minimumByteCount,
              encryptedData.prefix(magic.count) == magic
        else {
            throw EncryptedThemeAssetError.invalidAsset
        }

        var offset = magic.count
        let salt = encryptedData.subdata(in: offset..<(offset + saltByteCount))
        offset += saltByteCount
        let nonceData = encryptedData.subdata(in: offset..<(offset + nonceByteCount))
        offset += nonceByteCount

        let sealedData = encryptedData.subdata(in: offset..<encryptedData.count)
        guard sealedData.count > tagByteCount else {
            throw EncryptedThemeAssetError.invalidAsset
        }

        let ciphertext = sealedData.prefix(sealedData.count - tagByteCount)
        let tag = sealedData.suffix(tagByteCount)
        let symmetricKey = try deriveKey(from: keyText, salt: salt)
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.SealedBox(
            nonce: nonce,
            ciphertext: Data(ciphertext),
            tag: Data(tag)
        )

        return try AES.GCM.open(sealedBox, using: symmetricKey, authenticating: associatedData)
    }

    private static func deriveKey(from keyText: String, salt: Data) throws -> SymmetricKey {
        let passwordData = Data(keyText.utf8)
        guard !passwordData.isEmpty else {
            throw EncryptedThemeAssetError.emptyKey
        }

        var derivedKey = Data(repeating: 0, count: keyByteCount)
        let status = passwordData.withUnsafeBytes { passwordBuffer in
            salt.withUnsafeBytes { saltBuffer in
                derivedKey.withUnsafeMutableBytes { derivedBuffer in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBuffer.bindMemory(to: Int8.self).baseAddress,
                        passwordData.count,
                        saltBuffer.bindMemory(to: UInt8.self).baseAddress,
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        keyDerivationRounds,
                        derivedBuffer.bindMemory(to: UInt8.self).baseAddress,
                        keyByteCount
                    )
                }
            }
        }

        guard status == kCCSuccess else {
            throw EncryptedThemeAssetError.keyDerivationFailed
        }

        return SymmetricKey(data: derivedKey)
    }
}

enum EncryptedThemeAssetError: Error {
    case emptyKey
    case invalidAsset
    case keyDerivationFailed
}
